File: Tests/VIPSImageCGImageTests.m
//
//  VIPSImageCGImageTests.m
//  VIPSKitTests
//
//  Tests for VIPSImage+CGImage: CoreGraphics integration methods
//

#import "VIPSImageTestCase.h"

@interface VIPSImageCGImageTests : VIPSImageTestCase
@end

@implementation VIPSImageCGImageTests

#pragma mark - Basic CGImage Tests

- (void)testCreateCGImage {
    VIPSImage *image = [self createTestImageWithWidth:100 height:100];

    NSError *error = nil;
    CGImageRef cgImage = [image createCGImageWithError:&error];

    XCTAssertTrue(cgImage != NULL, @"Should create CGImage");
    XCTAssertNil(error, @"Should not have error: %@", error);
    XCTAssertEqual(CGImageGetWidth(cgImage), 100, @"CGImage width should match");
    XCTAssertEqual(CGImageGetHeight(cgImage), 100, @"CGImage height should match");

    if (cgImage) {
        CGImageRelease(cgImage);
    }
}

- (void)testCreateCGImageWithAlpha {
    VIPSImage *image = [self createTestImageWithWidth:100 height:100 bands:4];

    NSError *error = nil;
    CGImageRef cgImage = [image createCGImageWithError:&error];

    XCTAssertTrue(cgImage != NULL, @"Should create CGImage from RGBA");
    XCTAssertNil(error, @"Should not have error: %@", error);

    if (cgImage) {
        // Check that it has alpha
        CGImageAlphaInfo alphaInfo = CGImageGetAlphaInfo(cgImage);
        XCTAssertNotEqual(alphaInfo, kCGImageAlphaNone, @"Should have alpha channel");
        CGImageRelease(cgImage);
    }
}

- (void)testCreateCGImageGrayscale {
    VIPSImage *image = [self createTestImageWithWidth:100 height:100];

    NSError *error = nil;
    VIPSImage *gray = [image grayscaleWithError:&error];
    XCTAssertNotNil(gray, @"Should convert to grayscale");

    CGImageRef cgImage = [gray createCGImageWithError:&error];
    XCTAssertTrue(cgImage != NULL, @"Should create CGImage from grayscale");
    XCTAssertNil(error, @"Should not have error: %@", error);

    if (cgImage) {
        CGImageRelease(cgImage);
    }
}

#pragma mark - Thumbnail CGImage Tests

- (void)testCreateThumbnailCGImageFromLargeFile {
    // Test direct decode to thumbnail CGImage
    NSString *path = [self pathForTestResource:@"superman.jpg"];
    if (!path) {
        XCTSkip(@"superman.jpg not found in test resources");
        return;
    }

    NSError *error = nil;
    CGImageRef cgImage = [VIPSImage createThumbnailFromFile:path width:200 height:200 error:&error];

    XCTAssertTrue(cgImage != NULL, @"Should create thumbnail CGImage: %@", error);
    XCTAssertNil(error, @"Should not have error");

    if (cgImage) {
        NSInteger width = CGImageGetWidth(cgImage);
        NSInteger height = CGImageGetHeight(cgImage);

        XCTAssertLessThanOrEqual(width, 200, @"Width should fit within bounds");
        XCTAssertLessThanOrEqual(height, 200, @"Height should fit within bounds");

        CGImageRelease(cgImage);
    }
}

- (void)testCreateThumbnailCGImageSmall {
    NSString *path = [self pathForTestResource:@"superman.jpg"];
    if (!path) {
        XCTSkip(@"superman.jpg not found in test resources");
        return;
    }

    NSError *error = nil;
    CGImageRef cgImage = [VIPSImage createThumbnailFromFile:path width:50 height:50 error:&error];

    XCTAssertTrue(cgImage != NULL, @"Should create small thumbnail CGImage");

    if (cgImage) {
        XCTAssertLessThanOrEqual((NSInteger)CGImageGetWidth(cgImage), 50, @"Width should fit");
        XCTAssertLessThanOrEqual((NSInteger)CGImageGetHeight(cgImage), 50, @"Height should fit");
        CGImageRelease(cgImage);
    }
}

#pragma mark - Color Space Tests

- (void)testCreateCGImageWithColorSpace {
    VIPSImage *image = [self createTestImageWithWidth:100 height:100];

    NSError *error = nil;
    CGColorSpaceRef colorSpace = CGColorSpaceCreateDeviceRGB();
    CGImageRef cgImage = [image createCGImageWithColorSpace:colorSpace error:&error];

    XCTAssertTrue(cgImage != NULL, @"Should create CGImage with custom color space");
    XCTAssertNil(error, @"Should not have error: %@", error);

    CGColorSpaceRelease(colorSpace);
    if (cgImage) {
        CGImageRelease(cgImage);
    }
}

#pragma mark - Edge Cases

- (void)testCreateCGImageFromProcessedImage {
    VIPSImage *image = [self createTestImageWithWidth:200 height:200];
    NSError *error = nil;

    // Apply some processing
    VIPSImage *processed = [image resizeWithScale:0.5 error:&error];
    XCTAssertNotNil(processed, @"Should resize");

    processed = [processed adjustContrast:1.2 error:&error];
    XCTAssertNotNil(processed, @"Should adjust contrast");

    // Create CGImage from processed result
    CGImageRef cgImage = [processed createCGImageWithError:&error];
    XCTAssertTrue(cgImage != NULL, @"Should create CGImage from processed image");
    XCTAssertEqual(CGImageGetWidth(cgImage), 100, @"Width should match processed size");

    if (cgImage) {
        CGImageRelease(cgImage);
    }
}

- (void)testCreateCGImageFromEdgeDetection {
    VIPSImage *image = [self createTestImageWithWidth:100 height:100];
    NSError *error = nil;

    // Test with Sobel (single-band output)
    VIPSImage *sobel = [image sobelWithError:&error];
    XCTAssertNotNil(sobel, @"Should apply Sobel");

    CGImageRef sobelCG = [sobel createCGImageWithError:&error];
    XCTAssertTrue(sobelCG != NULL, @"Should create CGImage from Sobel: %@", error);
    if (sobelCG) CGImageRelease(sobelCG);

    // Test with Canny (single-band output)
    VIPSImage *canny = [image cannyWithSigma:1.4 error:&error];
    XCTAssertNotNil(canny, @"Should apply Canny");

    CGImageRef cannyCG = [canny createCGImageWithError:&error];
    XCTAssertTrue(cannyCG != NULL, @"Should create CGImage from Canny: %@", error);
    if (cannyCG) CGImageRelease(cannyCG);
}

@end

File: Tests/VIPSImageCGImageTests.swift
import XCTest
import CoreGraphics
import vips
import CVIPS
@testable import VIPSKit

final class VIPSImageCGImageTests: VIPSImageTestCase {

    // MARK: - Basic CGImage Creation

    func testCreateCGImage() throws {
        let image = createTestImage(width: 100, height: 100)
        let cgImage = try image.cgImage
        XCTAssertEqual(cgImage.width, 100)
        XCTAssertEqual(cgImage.height, 100)
    }

    func testCreateCGImageWithAlpha() throws {
        let image = createTestImage(width: 100, height: 100, bands: 4)
        let cgImage = try image.cgImage
        XCTAssertEqual(cgImage.width, 100)
        XCTAssertEqual(cgImage.height, 100)
    }

    func testCreateCGImageGrayscale() throws {
        let image = createTestImage(width: 100, height: 100)
        let gray = try image.grayscaled()
        let cgImage = try gray.cgImage
        XCTAssertEqual(cgImage.width, 100)
        XCTAssertEqual(cgImage.height, 100)
    }

    // MARK: - CGImage Property Verification

    func testCGImagePropertiesRGB() throws {
        let image = createTestImage(width: 50, height: 50)
        let cgImage = try image.cgImage
        XCTAssertEqual(cgImage.bitsPerComponent, 8)
        XCTAssertEqual(cgImage.bitsPerPixel, 24)
        XCTAssertEqual(cgImage.alphaInfo, .none)
        XCTAssertEqual(cgImage.colorSpace?.model, .rgb)
    }

    func testCGImagePropertiesRGBA() throws {
        let image = createTestImage(width: 50, height: 50, bands: 4)
        let cgImage = try image.cgImage
        XCTAssertEqual(cgImage.bitsPerComponent, 8)
        XCTAssertEqual(cgImage.bitsPerPixel, 32)
        XCTAssertEqual(cgImage.alphaInfo, .last)
        XCTAssertEqual(cgImage.colorSpace?.model, .rgb)
    }

    func testCGImagePropertiesGrayscale() throws {
        let image = createTestImage(width: 50, height: 50)
        let gray = try image.grayscaled()
        let cgImage = try gray.cgImage
        XCTAssertEqual(cgImage.bitsPerComponent, 8)
        XCTAssertEqual(cgImage.bitsPerPixel, 8)
        XCTAssertEqual(cgImage.alphaInfo, .none)
        XCTAssertEqual(cgImage.colorSpace?.model, .monochrome)
    }

    // MARK: - Band Count Edge Cases

    func testCGImageTwoBand() throws {
        // 2-band (grayscale + alpha) with B_W interpretation triggers the
        // bands == 2 conversion path in cgImage
        let gray = try createTestImage(width: 50, height: 50).grayscaled()
        let grayAlpha = try gray.addingAlpha()
        XCTAssertEqual(grayAlpha.bands, 2)
        let cgImage = try grayAlpha.cgImage
        XCTAssertEqual(cgImage.width, 50)
        XCTAssertEqual(cgImage.height, 50)
        XCTAssertEqual(cgImage.colorSpace?.model, .rgb)
    }

    func testCGImageFiveBand() throws {
        // >4 bands triggers the extract-first-4-bands path
        let rgba = createTestImage(width: 50, height: 50, bands: 4)
        let fiveBand = try rgba.appendBand(constant: 128.0)
        XCTAssertEqual(fiveBand.bands, 5)
        let cgImage = try fiveBand.cgImage
        XCTAssertEqual(cgImage.width, 50)
        XCTAssertEqual(cgImage.height, 50)
        XCTAssertEqual(cgImage.bitsPerPixel, 32)
        XCTAssertEqual(cgImage.alphaInfo, .last)
    }

    // MARK: - Colorspace Conversion

    func testCGImageFromLabColorspace() throws {
        // Convert sRGB → Lab, then verify cgImage converts back to sRGB
        let rgb = createTestImage(width: 50, height: 50)
        var labPtr: UnsafeMutablePointer<VipsImage>?
        guard cvips_colourspace(rgb.pointer, &labPtr, VIPS_INTERPRETATION_LAB) == 0,
              let labPtr else {
            XCTFail("Failed to convert to Lab colorspace")
            return
        }
        let labImage = VIPSImage(pointer: labPtr)
        XCTAssertEqual(vips_image_get_interpretation(labImage.pointer), VIPS_INTERPRETATION_LAB)

        let cgImage = try labImage.cgImage
        XCTAssertEqual(cgImage.width, 50)
        XCTAssertEqual(cgImage.height, 50)
        XCTAssertEqual(cgImage.colorSpace?.model, .rgb)
    }

    // MARK: - Format Casting

    func testCGImageFromFloatFormat() throws {
        // adjustBrightness produces FLOAT format, testing the cast-to-UCHAR path
        let image = createTestImage(width: 50, height: 50)
        let bright = try image.adjustBrightness(0.1)
        let cgImage = try bright.cgImage
        XCTAssertEqual(cgImage.width, 50)
        XCTAssertEqual(cgImage.height, 50)
        XCTAssertEqual(cgImage.bitsPerComponent, 8)
    }

    // MARK: - Thumbnail CGImage (Sync)

    func testThumbnailCGImage() throws {
        guard let path = pathForTestResource("superman.jpg") else {
            XCTFail("Test resource not found")
            return
        }
        let cgImage = try VIPSImage.thumbnailCGImage(fromFile: path, width: 100, height: 100)
        XCTAssertLessThanOrEqual(cgImage.width, 100)
        XCTAssertLessThanOrEqual(cgImage.height, 100)
    }

    func testThumbnailCGImageCGSize() throws {
        guard let path = pathForTestResource("superman.jpg") else {
            XCTFail("Test resource not found")
            return
        }
        let cgImage = try VIPSImage.thumbnailCGImage(fromFile: path, size: CGSize(width: 80, height: 80))
        XCTAssertLessThanOrEqual(cgImage.width, 80)
        XCTAssertLessThanOrEqual(cgImage.height, 80)
    }

    func testThumbnailCGImageInvalidPath() throws {
        XCTAssertThrowsError(
            try VIPSImage.thumbnailCGImage(fromFile: "/nonexistent/path.jpg", width: 100, height: 100)
        )
    }

    func testThumbnailPreservesAspectRatio() throws {
        guard let path = pathForTestResource("superman.jpg") else {
            XCTFail("Test resource not found")
            return
        }
        let original = try VIPSImage(contentsOfFile: path)
        let aspectRatio = Double(original.width) / Double(original.height)
        let cgImage = try VIPSImage.thumbnailCGImage(fromFile: path, width: 200, height: 100)
        let thumbAspect = Double(cgImage.width) / Double(cgImage.height)
        XCTAssertEqual(thumbAspect, aspectRatio, accuracy: 0.1)
    }

    func testThumbnailCGImageFromPNG() throws {
        guard let path = pathForTestResource("test-rgb.png") else {
            XCTFail("Test resource not found")
            return
        }
        let cgImage = try VIPSImage.thumbnailCGImage(fromFile: path, width: 50, height: 50)
        XCTAssertLessThanOrEqual(cgImage.width, 50)
        XCTAssertLessThanOrEqual(cgImage.height, 50)
        XCTAssertGreaterThan(cgImage.width, 0)
        XCTAssertGreaterThan(cgImage.height, 0)
    }

    func testThumbnailCGImageFromGrayscaleJPEG() throws {
        guard let path = pathForTestResource("grayscale.jpg") else {
            XCTFail("Test resource not found")
            return
        }
        let cgImage = try VIPSImage.thumbnailCGImage(fromFile: path, width: 50, height: 50)
        XCTAssertLessThanOrEqual(cgImage.width, 50)
        XCTAssertLessThanOrEqual(cgImage.height, 50)
    }

    // MARK: - CGImage from Real Files

    func testCGImageFromRGBAPNG() throws {
        guard let path = pathForTestResource("test-rgba.png") else {
            XCTFail("Test resource not found")
            return
        }
        let image = try VIPSImage(contentsOfFile: path)
        XCTAssertTrue(image.hasAlpha)
        let cgImage = try image.cgImage
        XCTAssertEqual(cgImage.width, image.width)
        XCTAssertEqual(cgImage.height, image.height)
        XCTAssertEqual(cgImage.alphaInfo, .last)
    }

    func testCGImageFromGrayscaleJPEG() throws {
        guard let path = pathForTestResource("grayscale.jpg") else {
            XCTFail("Test resource not found")
            return
        }
        let image = try VIPSImage(contentsOfFile: path)
        let cgImage = try image.cgImage
        XCTAssertEqual(cgImage.width, image.width)
        XCTAssertEqual(cgImage.height, image.height)
        XCTAssertEqual(cgImage.colorSpace?.model, .monochrome)
    }

    // MARK: - Pipeline Tests

    func testCreateCGImageFromProcessed() throws {
        let image = createTestImage(width: 200, height: 200)
        let resized = try image.resizeToFit(width: 50, height: 50)
        let cgImage = try resized.cgImage
        XCTAssertLessThanOrEqual(cgImage.width, 50)
        XCTAssertLessThanOrEqual(cgImage.height, 50)
    }

    func testCreateCGImageFromEdges() throws {
        let image = createTestImage(width: 100, height: 100)
        let edges = try image.sobel()
        let cgImage = try edges.cgImage
        XCTAssertEqual(cgImage.width, 100)
        XCTAssertEqual(cgImage.bitsPerComponent, 8)
    }

    // MARK: - Async

    func testAsyncMakeCGImage() async throws {
        let image = createTestImage(width: 100, height: 100)
        let cgImage = try await image.makeCGImage()
        XCTAssertEqual(cgImage.width, 100)
        XCTAssertEqual(cgImage.height, 100)
    }

    func testAsyncThumbnailCGImage() async throws {
        guard let path = pathForTestResource("superman.jpg") else {
            XCTFail("Test resource not found")
            return
        }
        let cgImage = try await VIPSImage.thumbnailCGImage(fromFile: path, width: 100, height: 100)
        XCTAssertLessThanOrEqual(cgImage.width, 100)
        XCTAssertLessThanOrEqual(cgImage.height, 100)
    }

    func testAsyncThumbnailCGImageCGSize() async throws {
        guard let path = pathForTestResource("superman.jpg") else {
            XCTFail("Test resource not found")
            return
        }
        let cgImage = try await VIPSImage.thumbnailCGImage(fromFile: path, size: CGSize(width: 80, height: 80))
        XCTAssertLessThanOrEqual(cgImage.width, 80)
        XCTAssertLessThanOrEqual(cgImage.height, 80)
    }
}
